scheduler/batch.go

```diff
@@ -0,0 +1,133 @@
+package scheduler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/apache/arrow/go/v15/arrow/array"
+	"github.com/apache/arrow/go/v15/arrow/memory"
+	"github.com/cloudquery/plugin-sdk/v4/message"
+	"github.com/cloudquery/plugin-sdk/v4/scalar"
+	"github.com/cloudquery/plugin-sdk/v4/schema"
+	"github.com/cloudquery/plugin-sdk/v4/writers"
+)
+
+type batcher struct {
+	res chan<- message.SyncMessage
+
+	size    int
+	timeout time.Duration
+
+	// using sync primitives by value here implies that batcher is to be used by pointer only
+	workers sync.Map // k = table name, v = *worker
+	wg      sync.WaitGroup
+}
+
+type worker struct {
+	ch      chan *schema.Resource
+	flush   chan chan struct{}
+	rows    schema.Resources
+	builder *array.RecordBuilder // we can reuse that
+	res     chan<- message.SyncMessage
+}
+
+func (w *worker) sync() {
+	if len(w.rows) == 0 {
+		return
+	}
+
+	// at least 1 row
+	w.builder.Reserve(len(w.rows)) // prealloc
+
+	for _, row := range w.rows {
+		scalar.AppendToRecordBuilder(w.builder, row.GetValues())
+	}
+
+	w.res <- &message.SyncInsert{Record: w.builder.NewRecord()}
+}
+
+func (w *worker) work(ctx context.Context, size int, timeout time.Duration) {
+	ticker := writers.NewTicker(timeout)
+	defer ticker.Stop()
+	flush := func() {
+		w.sync()
+		ticker.Reset(timeout)
+		w.rows = w.rows[:0]
+	}
+	for {
+		select {
+		case r, ok := <-w.ch:
+			if !ok {
+				w.sync()
+				return
+			}
+
+			w.rows = append(w.rows, r)
+			if size > 0 && len(w.rows) == size {
+				flush()
+			}
+
+		case <-ticker.Chan():
+			flush()
+
+		case done := <-w.flush:
+			flush()
+			close(done)
+
+		case <-ctx.Done():
+			// this means the request was cancelled
+			return // after this NO other call will succeed
```

Comment:
Not flushing here. Intentional?

Reply:
yep, as ctx was cancelled (copy-paste from batchwriter)